src/ship_cmd.cpp

```diff
@@ -574,6 +594,21 @@ static void ShipController(Ship *v)
 		return;
 	}
 
+	/* Moving up/down through lock */
+	int dz = ShipMoveUpDownOnLock(v);
+	if (dz != 0) {
+		if (v->cur_speed != 0) {
+			v->cur_speed = 0;
+			SetWindowWidgetDirty(WC_VEHICLE_VIEW, v->index, WID_VV_START_STOP);
+		}
```

Comment:
If these lines could be moved to a separate function, it would make future development of the ship controller easier.
Can these be included in a modified `bool ShipMoveUpDownOnLock(vehicle)` function?

Reply:
The intention was that the function was just a test.
I also note that there are other places where v->cur_speed is simply reset to 0 without the invalidation, so it may be fine to just set it to 0 here, or indeed create a separate function that can be used elsewhere.

Reply:
I mean moving all the new code on ship_cmd into ShipMoveUpDownOnLock so that in the ShipController only needs this single addition:
`if (ShipMoveUpDownOnLock(vehicle)) return;`
I checked CheckShipLeaveDepot and it does check whether it can leave the depot and also includes the lines for leaving the depot.
It is not a big improvement, but I would appreciate it.